tianshou/env/vecenv.py

```diff
@@ -180,6 +180,93 @@ def close(self) -> List[Any]:
         return result
 
 
+class AsyncVectorEnv(SubprocVectorEnv):
+    """Vectorized asynchronous environment wrapper based on subprocess.
+    .. seealso::
+
+        Please refer to :class:`~tianshou.env.BaseVectorEnv` for more detailed
+        explanation.
+    """
+    def __init__(self, env_fns: List[Callable[[], gym.Env]],
+                 wait_num: Optional[int] = None) -> None:
+        """
+        :param wait_num: used in asynchronous simulation if the time cost of
+            ``env.step`` varies with time and synchronously waiting for all
+            environments to finish a step is time-wasting. In that case, we
+            can return when ``wait_num`` environments finish a step and keep
+            on simulation in these environments. If ``None``, asynchronous
+            simulation is disabled; else, ``1 <= wait_num <= env_num``.
+        """
+        super().__init__(env_fns)
+        self.wait_num = wait_num or len(env_fns)
+        assert 1 <= self.wait_num <= len(env_fns), \
+            f'wait_num should be in [1, {len(env_fns)}], but got {wait_num}'
+        self.waiting_conn = []
+        self.waiting_id = []
+
+    def reset(self, id: Optional[Union[int, List[int]]] = None) -> np.ndarray:
+        if id is None:
+            id = range(self.env_num)
+        elif np.isscalar(id):
+            id = [id]
+        # reset envs are not waiting anymore
+        rest_envs = [(i, conn) for i, conn in
+                     zip(self.waiting_id, self.waiting_conn) if i not in id]
+        if not rest_envs:
+            self.waiting_id, self.waiting_conn = [], []
+        else:
+            self.waiting_id, self.waiting_conn = zip(*rest_envs)
+        return super().reset(id)
+
+    def render(self, **kwargs) -> List[Any]:
+        if len(self.waiting_id) > 0:
+            raise RuntimeError(
+                f"environments {self.waiting_id} are still "
+                f"stepping, cannot render now")
+        return super().render(**kwargs)
+
+    def close(self) -> List[Any]:
+        if self.closed:
+            return []
+        # finish remaining steps, and close
+        self.step(None)
+        return super().close()
+
+    def step(self,
+             action: Optional[np.ndarray],
+             id: Optional[Union[int, List[int]]] = None
+             ) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
+        """
+        Provide the given action to the environments specified by the id.
+        If action is None, fetch unfinished step() calls instead.
+        """
+        if id is not None:
+            raise ValueError("cannot specify the id of environments"
+                             " during step() for AsyncVectorEnv")
+        if action is not None:
+            assert len(action) + len(self.waiting_id) <= self.env_num
```

Comment:
I'm afraid it is not working like that. the action is specific to a given environment and cannot be executed by all of them indifferently. Indeed, the action has been computed based on a given observation of the environment, thus this exact same environment must be used to perform a step using this action.

Reply:
Nice analysis! I should use a queue to enforce the correspondence between the action (actually the observation to compute the action) and the environment.

Reply:
@duburcqa  please take a look at ef09ac2 😊

Reply:
Much better now !